File: agent/tools/dexscreener.py
"""DexScreener API tool - fetch DEX liquidity, volume, and trading pair data."""
import httpx
from typing import Optional


DEXSCREENER_BASE = "https://api.dexscreener.com/latest/dex"


async def search_pairs(query: str) -> list[dict]:
    """Search for trading pairs by token name or address."""
    async with httpx.AsyncClient(timeout=10) as client:
        resp = await client.get(f"{DEXSCREENER_BASE}/search", params={"q": query})
        resp.raise_for_status()
        return resp.json().get("pairs", []) or []


def parse_top_pairs(pairs: list[dict], max_pairs: int = 3) -> list[dict]:
    """Extract key liquidity/volume info from the top pairs (by liquidity)."""
    sorted_pairs = sorted(
        pairs,
        key=lambda p: float(p.get("liquidity", {}).get("usd", 0) or 0),
        reverse=True,
    )

    result = []
    for pair in sorted_pairs[:max_pairs]:
        txns = pair.get("txns", {}).get("h24", {})
        result.append({
            "chain": pair.get("chainId"),
            "dex": pair.get("dexId"),
            "pair": pair.get("pairAddress"),
            "url": pair.get("url"),
            "price_usd": pair.get("priceUsd"),
            "liquidity_usd": pair.get("liquidity", {}).get("usd"),
            "volume_24h": pair.get("volume", {}).get("h24"),
            "price_change_1h_pct": pair.get("priceChange", {}).get("h1"),
            "price_change_24h_pct": pair.get("priceChange", {}).get("h24"),
            "buys_24h": txns.get("buys"),
            "sells_24h": txns.get("sells"),
            "fdv": pair.get("fdv"),
            "market_cap": pair.get("marketCap"),
            "created_at": pair.get("pairCreatedAt"),
        })
    return result


def assess_liquidity_risk(pairs: list[dict]) -> str:
    """Simple liquidity risk assessment."""
    if not pairs:
        return "⚠️ 未找到交易对，极高风险"

    top_liquidity = float(pairs[0].get("liquidity_usd") or 0)

    if top_liquidity < 50_000:
        return f"🔴 流动性极低 (${top_liquidity:,.0f})，极易被砸盘"
    elif top_liquidity < 500_000:
        return f"🟡 流动性偏低 (${top_liquidity:,.0f})，需谨慎"
    elif top_liquidity < 5_000_000:
        return f"🟢 流动性尚可 (${top_liquidity:,.0f})"
    else:
        return f"🟢 流动性充足 (${top_liquidity:,.0f})"

File: src/tools/dexscreener.ts
import type { DexPair } from "../types/dex"

const BASE = "https://api.dexscreener.com/latest/dex"

function asRecord(value: unknown): Record<string, unknown> {
  return value != null && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : {}
}

function asString(value: unknown): string {
  return typeof value === "string" ? value : ""
}

function asNumber(value: unknown): number | undefined {
  const numberValue = typeof value === "number" ? value : Number(value)
  return Number.isFinite(numberValue) ? numberValue : undefined
}

export async function searchPairs(query: string): Promise<unknown[]> {
  try {
    const params = new URLSearchParams({ q: query })
    const res = await fetch(`${BASE}/search?${params}`, {
      signal: AbortSignal.timeout(10_000),
    })
    if (!res.ok) return []

    const data = await res.json() as unknown
    const pairs = asRecord(data).pairs
    return Array.isArray(pairs) ? pairs : []
  } catch {
    return []
  }
}

export function parseTopPairs(pairs: unknown[], maxPairs = 3): DexPair[] {
  const sorted = [...pairs]
    .map(asRecord)
    .sort((a, b) => {
      const liqA = asNumber(asRecord(a.liquidity).usd) ?? 0
      const liqB = asNumber(asRecord(b.liquidity).usd) ?? 0
      return liqB - liqA
    })

  return sorted.slice(0, maxPairs).map((pair) => {
    const txns = asRecord(asRecord(pair.txns).h24)
    const liquidity = asRecord(pair.liquidity)
    const volume = asRecord(pair.volume)
    const priceChange = asRecord(pair.priceChange)

    return {
      chain: asString(pair.chainId),
      dex: asString(pair.dexId),
      pairAddress: asString(pair.pairAddress),
      url: asString(pair.url),
      priceUsd: asNumber(pair.priceUsd),
      liquidityUsd: asNumber(liquidity.usd),
      volume24h: asNumber(volume.h24),
      priceChange1hPct: asNumber(priceChange.h1),
      priceChange24hPct: asNumber(priceChange.h24),
      buys24h: asNumber(txns.buys),
      sells24h: asNumber(txns.sells),
      fdv: asNumber(pair.fdv),
      marketCap: asNumber(pair.marketCap),
      createdAt: asNumber(pair.pairCreatedAt),
    }
  })
}

export function assessLiquidityRisk(pairs: DexPair[]): string {
  if (pairs.length === 0) return "⚠️ 未找到交易对，极高风险"

  const topLiquidity = pairs[0].liquidityUsd ?? 0
  const formattedLiquidity = topLiquidity.toLocaleString("en", { maximumFractionDigits: 0 })

  if (topLiquidity < 50_000) return `🔴 流动性极低 ($${formattedLiquidity})，极易被砸盘`
  if (topLiquidity < 500_000) return `🟡 流动性偏低 ($${formattedLiquidity})，需谨慎`
  if (topLiquidity < 5_000_000) return `🟢 流动性尚可 ($${formattedLiquidity})`
  return `🟢 流动性充足 ($${formattedLiquidity})`
}
